src/PrestaShopBundle/Controller/Admin/Order/DeliveryController.php

```diff
@@ -0,0 +1,173 @@
+use Symfony\Component\HttpFoundation\File\UploadedFile;
+use Symfony\Component\HttpFoundation\JsonResponse;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpFoundation\ResponseHeaderBag;
+
+/**
+ * Admin controller for the Order Delivery
+ */
+class DeliveryController extends FrameworkBundleAdminController
+{
+    /**
+     * Main page for Delivery slips.
+     *
+     * @Template("@PrestaShop/Admin/Order/Delivery/slip.html.twig")
+     * @param Request $request
+     *
+     * @return array|RedirectResponse
+     */
+    public function slipAction(Request $request)
+    {
+        $legacyController = $request->attributes->get('_legacy_controller');
+        if (!in_array(
+            $this->authorizationLevel($legacyController),
+            [
+                PageVoter::LEVEL_READ,
+                PageVoter::LEVEL_UPDATE,
+                PageVoter::LEVEL_CREATE,
+                PageVoter::LEVEL_DELETE,
+            ]
+        )) {
+            return $this->redirectToDefaultPage();
+        }
+
+        /* @var $formHandler FormHandler */
+        $formHandler = $this->get('prestashop.adapter.order.delivery.slip.form_handler');
+        /* @var $form Form */
+        $form = $formHandler->getForm();
+
+        $form->handleRequest($request);
+        if ($form->isSubmitted() && $form->isValid()) {
+            $errors = $formHandler->save($form->getData());
+            if (empty($errors)) {
+                $this->addFlash(
+                    'success',
+                    $this->trans('Update successful', 'Admin.Notifications.Success')
+                );
+            } else {
+                $this->flashErrors($errors);
+            }
+
+            return $this->redirectToRoute('admin_order_delivery_slip');
+        }
+
+        return [
+            'form' => $form->createView(),
+            'help_link' => $this->generateSidebarLink($legacyController),
+            'layoutTitle' => $this->trans('Delivery Slips', 'Admin.NavigationMenu'),
+            'requireAddonsSearch' => false,
+            'requireBulkActions' => false,
+            'showContentHeader' => true,
+            'enableSidebar' => true,
+        ];
+    }
+
+    /**
+     * Delivery slips PDF generator
+     *
+     * @param Request $request
+     *
+     * @return RedirectResponse
+     */
+    public function generatePdfAction(Request $request)
+    {
+        $legacyController = $request->attributes->get('_legacy_controller');
+        if (!in_array(
+            $this->authorizationLevel($legacyController),
+            [
+                PageVoter::LEVEL_READ,
+                PageVoter::LEVEL_UPDATE,
+                PageVoter::LEVEL_CREATE,
+                PageVoter::LEVEL_DELETE,
+            ]
+        )) {
+            return $this->redirectToDefaultPage();
+        }
+
+        $formData = $request->request->get('form');
```

Comment:
You're not allowed to do validation in controllers, rely on form providers/form handlers instead.

More, this way you don't provide any way to alter the data and forms with hooks.

See https://github.com/PrestaShop/prestafony-project/blob/master/docs/migration-guide.md#modern-form-management

Reply:
if for some reason, you can't rely on form handlers and providers, you need to create a service an extract the validation outside of controller.